apps/src/templates/rubrics/LearningGoals.jsx

```diff
@@ -246,11 +267,15 @@ export default function LearningGoals({
             if (json.understanding >= 0 && json.understanding !== null) {
               understandingLevels.current[index] = json.understanding;
             }
+            if (index === currentLearningGoal) {
+              setDisplayFeedback(teacherFeedbacks.current[currentLearningGoal]);
+              setDisplayUnderstanding(
+                understandingLevels.current[currentLearningGoal]
+              );
+            }
```

Comment:
This fixes an issue where the feedback would not get set if you opened the rubric container too quickly since it would otherwise rely on it polling the feedback again... which it would not do until you switched learning goals.

This fixes a flaky UI test as well... it would unsuccessfully race this behavior around 50% of the time.

Reply:
Wow!

Reply:
I think we're both wrong because `currentLearningGoal` is state and not a ref... when it performs the fetch, `currentLearningGoal` is baked and bound as a particular value (whatever it happens to be when you call fetch). If you quickly flip through the carousel, it will queue a bunch of these fetches and they all race. Each one thinks it is the current one and ALL of them will update the display potentially showing the wrong value.

`currentLearningGoal` should be a ref such that this callback pulls its current value and the callback is dependent on the `current.value` of the ref.

I think.

Reply:
I went ahead and added a very shallow ref that also tracks the current learning goal so that we don't get that update race when you click through the carousel quickly.

Reply:
CC @cearachew and @Nokondi who looked at this more closely than I did

Reply:
For the context now lost to my update, I made the exact same fix as Ceara: https://github.com/code-dot-org/code-dot-org/blob/bdbc3c7793585240305a0ef187724f21f5a05445/apps/src/templates/rubrics/LearningGoals.jsx#L249-L254

But when the callback happens `currentLearningGoal` is bound to the fetch... if it changes after the fetch happens, it will not be reflected in the fetch `then` so multiple `fetch` calls might all update and re-render because _their own_ idea (outdated) of the `currentLearningGoal` will line up with some `index`.